File: packages/llm/src/port.ts
import type { z } from "zod";

/**
 * Provider-agnostic LLM contract. The internal implementation (Vercel AI SDK) is isolated in a single
 * adapter file, and the composer / evals / SemanticPort implementations depend only on this Port.
 */

/** A Zod schema, or a raw JSON Schema dynamically built by the registry. */
export type SchemaInput<T> = z.ZodType<T> | { jsonSchema: Record<string, unknown> };

/**
 * An optional split of `prompt` into a leading `cacheable` part and a trailing `rest` part, for
 * providers with explicit prompt caching (Anthropic's `cache_control`). **Invariant:
 * `cacheable + rest === prompt`** — callers must construct both from the same content that produces
 * `prompt`, never as independently-computed strings, so the invariant is expected to hold by construction
 * at every call site. It is still a caller contract rather than something this type can enforce, so a
 * cache-capable adapter re-checks it defensively at the point of use (`cacheable + rest !== prompt`) and
 * falls back to sending the plain `prompt` string when it does not hold, rather than trusting a caller-side
 * bug to send mismatched content to the model. Purely additive: FakeLlm / FixtureLlm (and any LlmPort that
 * does not implement caching) ignore this field and read `prompt` as before, so passing it changes nothing
 * unless the adapter is both cache-capable and opted in (see LlmConfig.promptCache in @kohaku-ui/llm's env).
 * `cacheable` is expected to be the part that stays byte-identical across a run of calls that share it
 * (e.g. the composer's repair loop re-issuing the same static prompt prefix with only the trailing
 * repair-feedback section changing), so a provider that keys its cache on exact prefix bytes can reuse
 * the compiled/attended prefix instead of reprocessing it on every call.
 */
export interface PromptParts {
  cacheable: string;
  rest: string;
}

/**
 * Adaptive Reasoning effort level (Claude 4.6+/5's replacement for fixed thinking-token budgets; see
 * `adapters/ai-sdk.ts`'s `resolveProviderOptions` for the per-provider wiring and which providers ignore
 * it). Five levels from least to most reasoning effort spent before answering.
 */
export type LlmEffort = "low" | "medium" | "high" | "xhigh" | "max";

export interface GenerateObjectRequest<T> {
  schema: SchemaInput<T>;
  schemaName?: string;
  system?: string;
  prompt: string;
  /** See `PromptParts`'s doc. Optional; omitted entirely when the caller has no natural cache boundary. */
  promptParts?: PromptParts;
  /** Default 0 (aids determinism; the actual guarantee is provided by the cache). */
  temperature?: number;
  maxOutputTokens?: number;
  abort?: AbortSignal;
  /**
   * Output-budget multiplier (≥1, default 1). Scales this call's timeout (KOHAKU_LLM_TIMEOUT_MS) and,
   * when maxOutputTokens is not explicitly given, the output-token limit (KOHAKU_LLM_MAX_OUTPUT_TOKENS)
   * to the environment setting × this multiplier. Prevents a call with an order-of-magnitude longer output
   * — such as L2 free-form generation (full HTML) versus the standard (small L1 JSON) — from sharing the same
   * single-call budget and running out on the time/token limit first (observed: local ollama's L2 generation
   * falls into ABORTED at the default 60s boundary). Invalid values (NaN / less than 1) are treated as 1.
   */
  outputBudgetFactor?: number;
  /**
   * How much the model should reason before answering (optional; a port that does not implement effort
   * control ignores this field — FakeLlm/FixtureLlm and every non-Anthropic-matching adapter path leave
   * behavior unchanged). Omitted entirely means "provider default." See `LlmEffort`'s doc and
   * `adapters/ai-sdk.ts`'s `resolveProviderOptions` for which providers this actually reaches and under
   * which provider-options field.
   */
  effort?: LlmEffort;
}

export interface LlmUsage {
  inputTokens: number;
  outputTokens: number;
}

export interface GenerateObjectResult<T> {
  object: T;
  usage: LlmUsage;
  model: string;
}

export interface GenerateTextRequest {
  system?: string;
  prompt: string;
  /** See `PromptParts`'s doc (same contract as GenerateObjectRequest.promptParts). */
  promptParts?: PromptParts;
  temperature?: number;
  maxOutputTokens?: number;
  abort?: AbortSignal;
  /** Output-budget multiplier (≥1, default 1). Same meaning as GenerateObjectRequest.outputBudgetFactor. */
  outputBudgetFactor?: number;
  /** Same meaning and same "ignored if unsupported" contract as GenerateObjectRequest.effort. */
  effort?: LlmEffort;
}

export interface LlmPort {
  readonly provider: string;
  readonly modelId: string;
  generateObject<T>(req: GenerateObjectRequest<T>): Promise<GenerateObjectResult<T>>;
  generateText(req: GenerateTextRequest): Promise<{ text: string; usage: LlmUsage }>;
  /**
   * Streaming structured-output generation (optional). Notifies the **cumulative partial object** during
   * generation via onPartial, while returning the same validated object as generateObject as the final result.
   *
   * Contract:
   * - partial is best-effort. It may never be called at all (e.g. under the prompt JSON fallback).
   * - Each onPartial is "the cumulative form up to that point" — the consumer must rebuild from scratch every
   *   time (so it stays safe even when a provider retry resends from the beginning). A throw from onPartial is
   *   swallowed (it does not break generation).
   * - On a port with no implementation, the caller must fall back to generateObject (behaviorally equivalent, no partials).
   */
  streamObject?<T>(
    req: GenerateObjectRequest<T> & { onPartial: (partial: unknown) => void },
  ): Promise<GenerateObjectResult<T>>;
}

export type LlmErrorCode = "CONFIG" | "INVALID_OUTPUT" | "PROVIDER" | "ABORTED";

export class LlmError extends Error {
  readonly code: LlmErrorCode;
  readonly provider?: string;
  readonly modelId?: string;

  constructor(
    code: LlmErrorCode,
    message: string,
    opts: { provider?: string; modelId?: string; cause?: unknown } = {},
  ) {
    super(message, { cause: opts.cause });
    this.name = "LlmError";
    this.code = code;
    this.provider = opts.provider;
    this.modelId = opts.modelId;
  }
}

export function isZodSchema<T>(schema: SchemaInput<T>): schema is z.ZodType<T> {
  return typeof (schema as { safeParse?: unknown }).safeParse === "function";
}

File: python/kohaku/src/kohaku/llm/port.py
"""Provider-agnostic LLM contract (port of TS port.ts).

The internal implementation (the OpenAI-compatible Chat Completions adapter) is isolated in
adapters/, and the composer / evals / SemanticPort implementations depend only on this Port.

TS's zod schemas (`z.ZodType<T>`) map, in Python, to base pydantic: `type[BaseModel]` /
`TypeAdapter` as a "validating schema", and `JsonSchema` as a raw JSON Schema dynamically
built by the registry (validation is the caller's responsibility). Field names change from
TS's camelCase to snake_case (this is the product-internal Python API, not the wire contract).
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from typing import Any, Literal, Protocol, TypeGuard

from pydantic import BaseModel, TypeAdapter, ValidationError

from .abort import AbortSignal

type OnPartial = Callable[[object], None]
"""The recipient of the **cumulative partial object** during generation. Called by stream_object."""

type LlmEffort = Literal["low", "medium", "high", "xhigh", "max"]
"""Adaptive Reasoning effort level (Claude 4.6+/5's replacement for fixed thinking-token budgets; port of
TS's `LlmEffort`, packages/llm/src/port.ts). Five levels from least to most reasoning effort spent before
answering. See `adapters/anthropic_native.py` / `adapters/openai_compat.py` for the per-provider wiring and
which providers ignore it (gemini has no matching option)."""


@dataclass(frozen=True)
class JsonSchema:
    """A raw JSON Schema dynamically built by the registry. Not validated (the caller's responsibility)."""

    json_schema: dict[str, Any]


type SchemaInput = type[BaseModel] | TypeAdapter[Any] | JsonSchema
"""Corresponds to zod's `z.ZodType<T> | { jsonSchema }`. A pydantic model / TypeAdapter / raw JSON Schema."""


@dataclass(frozen=True)
class PromptParts:
    """An optional split of `prompt` into a leading `cacheable` part and a trailing `rest` part, for
    providers with explicit prompt caching (Anthropic's `cache_control`). Port of TS's `PromptParts`
    (packages/llm/src/port.ts) — see its doc for the full contract.

    **Invariant: `cacheable + rest == prompt`** — callers must construct both from the same content that
    produces `prompt`, never as independently-computed strings, so the invariant holds by construction.
    Purely additive: FakeLlm / FixtureLlm-equivalents (and any LlmPort that does not implement caching)
    ignore this field and read `prompt` as before.
    """

    cacheable: str
    rest: str


class SchemaValidationError(Exception):
    """A schema validation failure. The caller translates it into an INVALID_OUTPUT LlmError."""


def is_validating_schema(schema: SchemaInput) -> bool:
    """Whether this is a validating schema (pydantic-derived). `JsonSchema` is treated as pass-through, so False."""
    return not isinstance(schema, JsonSchema)


def schema_to_json_schema(schema: SchemaInput) -> dict[str, Any]:
    """Convert a schema to a JSON Schema (dict) (used for structured output and prompt attachment)."""
    if isinstance(schema, JsonSchema):
        return schema.json_schema
    if isinstance(schema, TypeAdapter):
        return schema.json_schema()
    return schema.model_json_schema()


def validate_against_schema(schema: SchemaInput, value: object) -> Any:
    """Validate a value against the schema and return the validated value. Raises SchemaValidationError on failure.

    `JsonSchema` is returned as-is without validation (by contract the caller validates it separately).
    """
    if isinstance(schema, JsonSchema):
        return value
    try:
        if isinstance(schema, TypeAdapter):
            return schema.validate_python(value)
        return schema.model_validate(value)
    except ValidationError as err:
        raise SchemaValidationError(str(err)) from err


@dataclass(frozen=True)
class GenerateObjectRequest:
    # TS's GenerateObjectRequest<T> is generic over the schema type T, but the JsonSchema branch of
    # SchemaInput is untyped and cannot bind T, so the Python version is non-generic and its result is
    # object: Any (the caller can recover a static type by receiving it with a pydantic model if needed).
    # Intentional difference.
    schema: SchemaInput
    prompt: str
    schema_name: str | None = None
    system: str | None = None
    prompt_parts: PromptParts | None = None
    """See `PromptParts`'s doc. Optional; omitted entirely when the caller has no natural cache boundary."""
    temperature: float | None = None
    """Default 0 (aids determinism; the actual guarantee is provided by the cache)."""
    max_output_tokens: int | None = None
    abort: AbortSignal | None = None
    output_budget_factor: float | None = None
    """Output-budget multiplier (>=1, default 1). Scales this call's timeout and, when max_output_tokens
    is not explicitly given, the output-token limit, to the environment setting × this multiplier. Prevents
    a call with an order-of-magnitude longer output — such as L2 free-form generation (full HTML) versus the
    standard (small L1 JSON) — from sharing the same single-call budget and running out on the time/token
    limit first. Invalid values (non-finite / less than 1) are treated as 1."""
    effort: LlmEffort | None = None
    """How much the model should reason before answering (optional; a port that does not implement effort
    control ignores this field — FakeLlm and every provider adapter with no matching option leave behavior
    unchanged). Omitted entirely means "provider default." See `LlmEffort`'s doc for which providers this
    actually reaches and under which wire field."""


@dataclass(frozen=True)
class LlmUsage:
    input_tokens: int
    output_tokens: int


@dataclass(frozen=True)
class GenerateObjectResult:
    object: Any
    usage: LlmUsage
    model: str


@dataclass(frozen=True)
class GenerateTextRequest:
    prompt: str
    system: str | None = None
    prompt_parts: PromptParts | None = None
    """See `PromptParts`'s doc (same contract as GenerateObjectRequest.prompt_parts)."""
    temperature: float | None = None
    max_output_tokens: int | None = None
    abort: AbortSignal | None = None
    output_budget_factor: float | None = None
    """Output-budget multiplier (>=1, default 1). Same meaning as GenerateObjectRequest.output_budget_factor."""
    effort: LlmEffort | None = None
    """Same meaning and same "ignored if unsupported" contract as GenerateObjectRequest.effort."""


@dataclass(frozen=True)
class GenerateTextResult:
    text: str
    usage: LlmUsage


class LlmPort(Protocol):
    provider: str
    model_id: str

    async def generate_object(self, req: GenerateObjectRequest) -> GenerateObjectResult: ...

    async def generate_text(self, req: GenerateTextRequest) -> GenerateTextResult: ...


class StreamingLlmPort(LlmPort, Protocol):
    """An LlmPort that supports streaming structured-output generation (equivalent to TS's optional `streamObject`).

    Notifies the **cumulative partial object** during generation via on_partial, while returning the final
    result as the same validated object as generate_object. Contract:

    - partial is best-effort. It may never be called at all (e.g. under the prompt JSON fallback).
    - Each on_partial is "the cumulative form up to that point" — the consumer must rebuild from scratch every
      time (so it stays safe even when a provider retry resends from the beginning). A throw from on_partial is
      swallowed (it does not break generation).
    - On a port with no implementation, the caller must fall back to generate_object (behaviorally equivalent, no partials).

    TS passes `req & { onPartial }` as an intersection type, but Python has no intersection type, so on_partial
    is split out as the second argument (intentional difference). An "optional method" also cannot be expressed
    in a Protocol, so it is modeled as a separate Protocol + the supports_streaming TypeGuard (callers fall back to generate_object).
    """

    async def stream_object(
        self, req: GenerateObjectRequest, on_partial: OnPartial
    ) -> GenerateObjectResult: ...


def supports_streaming(llm: LlmPort) -> TypeGuard[StreamingLlmPort]:
    """Whether the LlmPort implements stream_object (incremental streaming).

    If not implemented, the caller falls back to generate_object (behaviorally equivalent, no partials).
    """
    return callable(getattr(llm, "stream_object", None))


type LlmErrorCode = Literal["CONFIG", "INVALID_OUTPUT", "PROVIDER", "ABORTED"]


class LlmError(Exception):
    """A structured error with a code. The conformance / composer repair loops branch on the code."""

    def __init__(
        self,
        code: LlmErrorCode,
        message: str,
        *,
        provider: str | None = None,
        model_id: str | None = None,
        cause: BaseException | None = None,
    ) -> None:
        super().__init__(message)
        self.code: LlmErrorCode = code
        self.provider = provider
        self.model_id = model_id
        if cause is not None:
            self.__cause__ = cause
